### PR TITLE
Don't create empty holding for MPT issuer. Enforce MPToken creation by non-issuer invariant.

### PR DESCRIPTION
**Pull request by gregtatcam**
## High Level Overview of Change

This fixes [RIPD-3462.](https://ripplelabs.atlassian.net/issues/?filter=-1&selectedIssue=RIPD-3462)

### Context of Change

Current addEmptyHolding() creates MPToken for an issuer. This is incorrect since an issuer can't create MPToken. Updated version of addEmptyHolding() returns tesSUCCESS if an account is the issuer.

Updated ValidMPTIssuance class in InvariantCheck enforces MPToken creation by non-issuer invariant.

### CODE REVIEW

**Comment by Bronek** [comment]
Please see #5875

File: src/xrpld/app/tx/detail/InvariantCheck.cpp [hunk 1]
```diff
@@ -1493,6 +1498,15 @@ ValidMPTIssuance::finalize(
 {
     if (result == tesSUCCESS)
     {
+        if (mptCreatedByIssuer_)
+        {
+            JLOG(j.fatal())
+                << "Invariant failed: MPT authorize submitted by issuer "
+                   "succeeded but created mptokens";
+            return false;
```
Comment (a1q123456) [inline src/xrpld/app/tx/detail/InvariantCheck.cpp]: QQ: What's the difference between failing here and failing in `ValidMPTIssuance::visitEntry`

File: src/libxrpl/ledger/View.cpp [hunk 2]
```diff
@@ -1282,6 +1282,8 @@ addEmptyHolding(
         return tefINTERNAL;  // LCOV_EXCL_LINE
     if (view.peek(keylet::mptoken(mptID, accountID)))
         return tecDUPLICATE;
+    if (accountID == mptIssue.getIssuer())
+        return tesSUCCESS;
```
Comment (shawnxie999) [inline src/libxrpl/ledger/View.cpp:1472]: is `tesSuccess` really the correct code to return here? From my intuition, an issuer should not be allowed to add an empty holding, therefore an error code should be returned instead - and the caller of `addEmptyHolding` should be responsible for checking that the accountID is not the issuer.

File: src/xrpld/app/tx/detail/InvariantCheck.cpp (hunk 1 above)
Comment (shawnxie999) [reply src/xrpld/app/tx/detail/InvariantCheck.cpp]: i don't think you can fail in visitEntry, it's a void function

Comment (shawnxie999) [inline src/xrpld/app/tx/detail/InvariantCheck.cpp]: i believe this needs to be amendment gated

Comment (a1q123456) [reply src/xrpld/app/tx/detail/InvariantCheck.cpp]: Ah yeah, thanks, Shawn!

File: src/libxrpl/ledger/View.cpp (hunk 2 above)
Comment (a1q123456) [reply src/libxrpl/ledger/View.cpp:1472]: I think this is the reason: https://github.com/XRPLF/rippled/pull/5875/files#r2426305173

**Review by shawnxie999 — CHANGES_REQUESTED** [review]
The invariant needs to be amendment gated (by SingleAssetVault?). Also, I don’t think we should silently return `tesSUCCESS` in `addEmptyHolding` when the account is the issuer - this could potentially lead to misuse in the future, where future callers of this function might overlook this edge case. I feel this condition should be explicitly handled by the caller instead. That said, this is subjective and not a blocker.

File: src/libxrpl/ledger/View.cpp (hunk 2 above)
Comment (shawnxie999) [reply src/libxrpl/ledger/View.cpp:1472]: although it's clean to handle this edge case in the function, i still think it might be something that's worth being explicitly handled by the caller for safety practice, as there could be downstream bugs if future callers aren't aware that this edge case is silently returned as tesSuccess. But again, we do have sufficient unit tests and invariants for so the likelihood of something like this going wrong is pretty low.

Comment (ximinez) [reply src/libxrpl/ledger/View.cpp:1472]: `temDST_IS_SRC` is the result returned by a `TrustSet` transaction if the account is the issuer. ~I think the situation is analogous enough that it would be appropriate here, too.~

This change does _not_ need to be amendment gated, because `addEmptyHolding` is only used with SAV and LP, which are already amendment gated.

File: src/libxrpl/ledger/View.cpp [hunk 3]
```diff
@@ -1226,8 +1226,8 @@ addEmptyHolding(
     Issue const& issue,
     beast::Journal journal)
 {
-    // Every account can hold XRP.
-    if (issue.native())
+    // Every account can hold XRP. An issuer can issue directly.
+    if (issue.native() || accountID == issue.getIssuer())
         return tesSUCCESS;
```
Comment (ximinez) [inline src/libxrpl/ledger/View.cpp:1411]: ~I think `temDST_IS_SRC` should be returned here, too.~

This change does _not_ need to be amendment gated, because `addEmptyHolding` is only used with SAV and LP, which are already amendment gated.

File: src/test/app/Loan_test.cpp [hunk 4]
```diff
@@ -2709,10 +2703,51 @@ class Loan_test : public beast::unit_test::suite
...
+        // Issuer should not create MPToken
+        BEAST_EXPECT(!env.le(keylet::mptoken(asset.issuanceID(), issuer)));
+        // Issuer "borrowed" 200, OutstandingAmount decreased by 200
+        BEAST_EXPECT(env.balance(issuer, asset) == asset(-issuerBalance + 200));
+        // Pay Loan
+        auto const loanKeylet = keylet::loan(broker.brokerID, 1);
+        env(pay(borrower, loanKeylet.key, asset(200)));
+        env.close();
+        // Issuer "re-payed" 200, OutstandingAmount increased by 200
+        BEAST_EXPECT(env.balance(issuer, asset) == asset(-issuerBalance));
```
Comment (ximinez) [inline src/test/app/Loan_test.cpp:2867]: Wait. You mean that by just avoiding creating the holding, everything else "just works"!?!?

File: src/xrpld/app/tx/detail/InvariantCheck.cpp (hunk 1 above)
Comment (ximinez) [reply src/xrpld/app/tx/detail/InvariantCheck.cpp]: > i believe this needs to be amendment gated

Agreed. I would have the gate check for either `SingleAssetVault` or `LendingProtocol` to be enabled.

**Comment by ximinez** [comment]
You know, on further thought. I think `tesSUCCESS` is the way to go.

My thinking is that if the caller needs to do something different to handle the case where the issuer is the relevant account, then `addEmptyHolding` should fail, and the callers should check the account, and do the alternate thing if it's the issuer.

But that's not the case. As your unit test demonstrates, nothing else needs to be done. Therefore, it is a complete solution to have `addEmptyHolding` return `tesSUCCESS` if the account is the issuer. In fact, that's better than checking in the caller, because then you don't have to worry about forgetting.

The one caveat is that this also has to hold true for `VaultCreate`. I suspect it does, because it's doing the same basic operations (and I stole a bunch of code from from it to start implementing these transactions). @Bronek Do you have any unit tests checking that scenario?

**Comment by Bronek** [comment]
> You know, on further thought. I think `tesSUCCESS` is the way to go.
> 
> My thinking is that if the caller needs to do something different to handle the case where the issuer is the relevant account, then `addEmptyHolding` should fail, and the callers should check the account, and do the alternate thing if it's the issuer.
> 
> But that's not the case. As your unit test demonstrates, nothing else needs to be done. Therefore, it is a complete solution to have `addEmptyHolding` return `tesSUCCESS` if the account is the issuer. In fact, that's better than checking in the caller, because then you don't have to worry about forgetting.
> 
> The one caveat is that this also has to hold true for `VaultCreate`. I suspect it does, because it's doing the same basic operations (and I stole a bunch of code from from it to start implementing these transactions). @Bronek Do you have any unit tests checking that scenario?

In `VaultCreate` we only call `addEmptyHolding` for the pseudo-account, which by definition cannot be an issuer of the assets stored in vault (logic: `canAddHolding` called inside `preclaim`, before we even created the pseudo-account). 

However this does impact `VaultWithdraw` where we call `addEmptyHolding` to ensure that there is somewhere to store the balance of the withdrawn assets. This is happening in https://github.com/XRPLF/rippled/blob/5e33ca56fd51b02a5490e8e156e2de65937cf88e/src/xrpld/app/tx/detail/VaultWithdraw.cpp#L296-L304

... and the relevant unit tests are in https://github.com/XRPLF/rippled/blob/5e33ca56fd51b02a5490e8e156e2de65937cf88e/src/test/app/Vault_test.cpp#L420-L439

I think with this change, the above displayed part of `VaultWithdraw` should be simplified.

**Comment by ximinez** [comment]
> I think with this change, the above displayed part of VaultWithdraw should be simplified.

Do you want to tack a commit making that change onto this PR?

**Comment by dangell7** [comment]
> You know, on further thought. I think `tesSUCCESS` is the way to go.
> 
> My thinking is that if the caller needs to do something different to handle the case where the issuer is the relevant account, then `addEmptyHolding` should fail, and the callers should check the account, and do the alternate thing if it's the issuer.
> 
> But that's not the case. As your unit test demonstrates, nothing else needs to be done. Therefore, it is a complete solution to have `addEmptyHolding` return `tesSUCCESS` if the account is the issuer. In fact, that's better than checking in the caller, because then you don't have to worry about forgetting.
> 
> The one caveat is that this also has to hold true for `VaultCreate`. I suspect it does, because it's doing the same basic operations (and I stole a bunch of code from from it to start implementing these transactions). @Bronek Do you have any unit tests checking that scenario?

See I actually go the other way a bit. If this was only supposed to be used in one place maybe return tesSUCCESS but if its agnostic then it probably should error I feel. However there is a GlobalFreeze check in there that makes it not agnostic because when locking tokens PayChan/Escrow/Offer/Subscriptions we handle GlobalFreeze differently based on who's receiving the funds.

I built this function the other day. I would really love to see us move towards this type of thing.

https://github.com/XRPLF/rippled/blob/0039fdb83fb379ee00c2e84e80fa5cd39e52ba7f/src/libxrpl/ledger/View.cpp#L2960

idk

**Comment by ximinez** [comment]
> See I actually go the other way a bit. If this was only supposed to be used in one place maybe return tesSUCCESS but if its agnostic then it probably should error I feel.

Consider that if you pass an `Asset` representing XRP to `addEmptyHolding`, it succeeds. Because there's nothing to do. The same applies if the account is the issuer.

**Review by ximinez — APPROVED** [review]

**Review by a1q123456 — APPROVED** [review]

**Review by shawnxie999 — APPROVED** [review]
